ddtrace/contrib/tornado/middlewares.py

```diff
@@ -0,0 +1,67 @@
+from tornado.web import Application
+from tornado.stack_context import StackContext
+
+from . import handlers
+from .stack_context import ContextManager
+from ...ext import AppTypes
+
+
+class TraceMiddleware(object):
+    """
+    Tornado middleware class that wraps a Tornado ``HTTPServer`` instance
+    so that the request_callback can be wrapped with a ``StackContext``
+    that uses the internal ``ContextManager``. This middleware creates
+    a root span for each request.
+    """
+    def __init__(self, http_server, tracer, service='tornado-web'):
+        """
+        Replace the default ``HTTPServer`` request callback with this
+        class instance that is callable. If the given request callback
+        is a Tornado ``Application``, all handlers are wrapped with
+        tracing methods.
+        """
+        self._http_server = http_server
+        self._tracer = tracer
+        self._service = service
+        # the default http_server callback must be preserved
+        self._request_callback = http_server.request_callback
```

Comment:
Thank you Ross for your details. Actually the work on Tornado isn't finished because there are cases like you said that should be handled. Also I'm defining what are the supported situations, given the fact that the tracer [has a new public API][1] that makes easier to instrument the code in asynchronous environment.

Will keep you updated.

[1]: https://github.com/DataDog/dd-trace-py/pull/172